src/cipher.rs

```diff
@@ -568,298 +569,184 @@ impl rustls_root_cert_store {
-            let crls_der = match crls_der{
-                Ok(vv) => vv,
-                Err(_) => return rustls_result::CertificateRevocationListParseError,
-            };
-
-            let client_cert_verifier = match client_cert_verifier_builder.take() {
-                None => {
-                    return AlreadyUsed;
-                },
-                Some(x) => x,
-            };
-
-            match client_cert_verifier.with_crls(crls_der) {
-                Ok(v) => client_cert_verifier_builder.replace(v),
-                Err(e) => return map_error(rustls::Error::InvalidCertRevocationList(e)),
-            };
-
-            rustls_result::Ok
-        }
-    }
-
-    /// Free a `rustls_allow_any_authenticated_client_builder` previously returned from
-    /// `rustls_allow_any_authenticated_client_builder_new`.
-    /// Calling with NULL is fine. Must not be called twice with the same value.
+impl rustls_client_cert_verifier {
+    /// Free a `rustls_client_cert_verifier` previously returned from
+    /// `rustls_client_cert_verifier_builder_build`. Calling with NULL is fine. Must not be
+    /// called twice with the same value.
     #[no_mangle]
-    pub extern "C" fn rustls_allow_any_authenticated_client_builder_free(
-        builder: *mut rustls_allow_any_authenticated_client_builder,
-    ) {
+    pub extern "C" fn rustls_client_cert_verifier_free(verifier: *mut rustls_client_cert_verifier) {
         ffi_panic_boundary! {
-            let store = try_box_from_ptr!(builder);
-            drop(store)
+            free_box(verifier);
         }
     }
 }
 
-/// A verifier of client certificates that requires all certificates to be
-/// trusted based on a given `rustls_root_cert_store`. Usable in building server
-/// configurations. Connections without such a client certificate will not
-/// be accepted.
-pub struct rustls_allow_any_authenticated_client_verifier {
+/// A client certificate verifier being constructed. A builder can be modified by,
+/// e.g. `rustls_web_pki_client_cert_verifier_builder_add_crl`. Once you're
+/// done configuring settings, call `rustls_web_pki_client_cert_verifier_builder_build`
+/// to turn it into a `rustls_client_cert_verifier`. This object is not safe
+/// for concurrent mutation.
+// TODO(@cpu): Add rustdoc link once available.
+pub struct rustls_web_pki_client_cert_verifier_builder {
+    // We use the opaque struct pattern to tell C about our types without
+    // telling them what's inside.
+    // https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs
     _private: [u8; 0],
 }
 
-impl Castable for rustls_allow_any_authenticated_client_verifier {
-    type Ownership = OwnershipArc;
-    type RustType = AllowAnyAuthenticatedClient;
+pub(crate) struct ClientCertVerifierBuilder {
+    roots: Arc<RootCertStore>,
+    crls: Vec<CertificateRevocationListDer<'static>>,
```

Comment:
Leaving a note here: I'm not sure that using a `'static` lifetime is the correct choice for this type. I'll give this more thought.

Reply:
I think I've convinced myself this is correct:

The `CertificateRevocationListDer` type is defined in `pki-types` as a newtype around `Der<'a>`. That in turn is a newtype around `DerInner<'a>` which is an enum with two variants:

```
#[derive(Clone, PartialEq)]
enum DerInner<'a> {
    #[cfg(feature = "alloc")]
    Owned(Vec<u8>),
    Borrowed(&'a [u8]),
}
```

The lifetime that gets carried through the newtypes is specific to the `Borrowed` variant, but we're using the `Owned(Vec<u8>)` variant throughout, matching what was previously `Vec<UnparsedCertRevocationList>`.

Using `'static` as the lifetime binds us to either sticking with the owned variant (that has no associated lifetime) as we've done here, or a `'static` lifetime for borrowed data we can treat equivalent to an owned type from a lifetime perspective.

Reply:
Yep, that's exactly right; sorry, I meant to reply on this thread earlier but I forgot!

Also worth noting there's a secondary but similar meaning for `'static` when used in a trait bound, which means "[the receiver can hold onto the type as long as they want](https://doc.rust-lang.org/rust-by-example/scope/lifetime/static_lifetime.html#trait-bound)"

Reply:
> Yep, that's exactly right

Great! Glad I came to the right conclusion :-)

> Also worth noting there's a secondary but similar meaning for 'static when used in a trait bound

Ah interesting! I think I've seen that but I don't know that I had a crisp understanding of what it meant.